src/java.desktop/windows/native/libawt/windows/awt_Debug.cpp

```diff
@@ -153,21 +153,20 @@ AwtDebugSupport::~AwtDebugSupport() {
 static jboolean isHeadless() {
     jmethodID headlessFn;
     JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
-    jclass graphicsEnvClass = env->FindClass(
-        "java/awt/GraphicsEnvironment");
+    // be on the safe side and avoid JNI warnings by calling ExceptionCheck
+    // an accumulated exception is not cleared
+    env->ExceptionCheck();
```

Comment:
Yes. However, it's "only" in the assertion callback that only exists in debug VMs. And an original exception isn't lost.

Reply:
Perhaps what we should do here is
if (ExceptionCheck) {
 ExceptionDescribe
 ExceptionClear
}
So someone can "see" the text of the original exception, and the debugging code is safe to make the calls it wants.

The alternatives are that the debugging code in the case of ExceptionCheck==TRUE just do what it takes to silence the JNI warnings , assuming that TRUE is never not a possibility, so no real problem, but I don't know see how we can be sure about that for ALL callers of this assert code. (BTW I wonder if the reason the current code didn't do as expected is because ExceptionCheck isn't doing what we expect, but I don't see how), or alternative number 2 is that the debug code simply bails in the face of a pending exception, ie
if (ExceptionCheck) {
   return;
}

Reply:
As far as I can see, the real problem is that `DWMIsCompositionEnabled` calls a Java method and does not check if an exception occurred. It should do it according to the JNI specification.

I can assume `initScreens(env)` does not call JNI methods, therefore no JNI warning is produced in the regular code flow where no assertions fail.

Reply:
> assuming that TRUE is never not a possibility, so no real problem, but I don't know see how we can be sure about that for ALL callers of this assert code.

We can never be sure about it, even though I tend to believe exceptions are rare.

Essentially, any upcall into Java followed by an assertion will lead to this JNI warning because the assertion handler also upcalls into Java.

Reply:
Yes, I'll take a look at that soon. Remember there are two issues behind all of this
(1) We don't have a real headless build for windows and this can cause failures of windows APIs and Java exceptions
(2) debug builds use asserts on failures of windows APIs 

(2) is partially addressed here
(1) is partially addressed in the other fix. I can't be sure its completely addressed

Reply:
I thought about it a bit more…

Since `Java_sun_awt_Win32GraphicsEnvironment_initDisplay` does not call other JNI functions after `DWMIsCompositionEnabled`, it is reasonable to leave a possible raised exception unhandled — it will handled by JVM after `initDisplay` returns.

Perhaps, it's better to handle the exception inside the assertion handler:
```cpp
if (env->ExceptionCheck()) {
    env->ExceptionDescribe();
    env->ExceptionClear();
}```

Or maybe not, if `ExceptionCheck()` is enough to silence the warning and the raised exception is still thrown after the control is returned to JVM.